src/Agent.Worker/AgentPluginManager.cs

```diff
@@ -0,0 +1,332 @@
+                }
+            }
+
+            // Load command plugin
+            foreach (var plugin in _commandPlugins)
+            {
+                IAgentCommandPlugin commandPlugin = null;
+                string typeName = $"{plugin.AgentPluginEntryPoint}, {plugin.AgentPluginAssembly}";
+                AssemblyLoadContext.Default.Resolving += ResolveAssembly;
+                try
+                {
+                    Trace.Info($"Load command plugin from '{typeName}'.");
+                    Type type = Type.GetType(typeName, throwOnError: true);
+                    commandPlugin = Activator.CreateInstance(type) as IAgentCommandPlugin;
+                }
+                finally
+                {
+                    AssemblyLoadContext.Default.Resolving -= ResolveAssembly;
+                }
+
+                ArgUtil.NotNull(commandPlugin, nameof(commandPlugin));
+                ArgUtil.NotNullOrEmpty(commandPlugin.Area, nameof(commandPlugin.Area));
+                ArgUtil.NotNullOrEmpty(commandPlugin.Event, nameof(commandPlugin.Event));
+                ArgUtil.NotNullOrEmpty(commandPlugin.DisplayName, nameof(commandPlugin.DisplayName));
+
+                if (!_supportedLoggingCommands.ContainsKey(commandPlugin.Area))
+                {
+                    _supportedLoggingCommands[commandPlugin.Area] = new Dictionary<string, AgentCommandPluginInfo>(StringComparer.OrdinalIgnoreCase);
+                }
+
+                Trace.Info($"Loaded command plugin to handle '##vso[{commandPlugin.Area}.{commandPlugin.Event}]'.");
+                _supportedLoggingCommands[commandPlugin.Area][commandPlugin.Event] = new AgentCommandPluginInfo() { CommandPluginTypeName = typeName, DisplayName = commandPlugin.DisplayName };
+            }
+        }
+
+        public async Task RunPluginTaskAsync(IExecutionContext context, string plugin, Dictionary<string, string> inputs, Dictionary<string, string> environment, string stage, EventHandler<ProcessDataReceivedEventArgs> outputHandler)
+        {
+            ArgUtil.NotNullOrEmpty(plugin, nameof(plugin));
+
+            // Resolve the working directory.
+            string workingDirectory = HostContext.GetDirectory(WellKnownDirectory.Bin);
+            ArgUtil.Directory(workingDirectory, nameof(workingDirectory));
+
+            // Agent.PluginHost
+            string file = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Bin), $"Agent.PluginHost{Util.IOUtil.ExeExtension}");
+            ArgUtil.File(file, $"Agent.PluginHost{Util.IOUtil.ExeExtension}");
+
+            // Agent.PluginHost's arguments
+            string arguments = $"task \"{plugin}\"";
+
+            // construct plugin context
+            AgentTaskPluginExecutionContext pluginContext = new AgentTaskPluginExecutionContext
+            {
+                Inputs = inputs,
+                Stage = stage,
+                Repositories = context.Repositories,
+                Endpoints = context.Endpoints
```

Comment:
is the host portion of the urls in the endpoints fixed up already? what about repositories, will they need to be fixed up too?

Reply:
everything in executioncontext should already be fixed up.

Reply:
in jobrunner, fix url for repository as well.

Reply:
fixed.